sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,6 +153,15 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         )
       )
 
+    // Push down limit into union
+    case Limit(exp, Union(left, right)) =>
+      Limit(exp,
+        Union(
+          CombineLimits(Limit(exp, left)),
+          CombineLimits(Limit(exp, right))
+        )
+      )
```

Comment:
After think it more, there may be a problem: If `left` or `right` is an operator that can push `Limit` down(currently there is no such operator, but we can't guarantee there won't be). Then every time you push down a `Limit` here, it will be pushed down further. Thus the `CombineLimits` can NOT detect that you have already pushed the `Limit` down, and keeps generating new `Limit` and pushing it down.

I think we should have a better way to detect whether we have pushed `Limit` down or not, or add some comments to say that this rule assumes the newly added `Limit` on top of `left` and `right` won't be removed by other optimization rules.